### PR TITLE
Feature/outer paths for project

### PR DESCRIPTION
**Pull request by udeeshagautam**
This PR enables having relative path outside of project folder. 

for a path like ..\something\Table2.sql and ..\Table3.sql the leaf node will always show at the project root level like following

image

### CODE REVIEW

File: extensions/sql-database-projects/src/common/constants.ts [hunk 1]
```diff
@@ -108,6 +108,7 @@ export const invalidDataSchemaProvider = localize('invalidDataSchemaProvider', "
 export const invalidDatabaseReference = localize('invalidDatabaseReference', "Invalid database reference in .sqlproj file");
 export const databaseSelectionRequired = localize('databaseSelectionRequired', "Database selection is required to import a project");
 export const databaseReferenceAlreadyExists = localize('databaseReferenceAlreadyExists', "A reference to this database already exists in this project");
+export const ousiderFolderPath = localize('outsideFolderPath', "Files are added to project file with absolute paths outside of projects drive. please make the path relative and add files from the same drive");
```
Comment (Benjin) [inline extensions/sql-database-projects/src/common/constants.ts:111]: A quick nit-read :)
* "adding to project with..." (avoiding "file" twice so close)
* project's drive (apostrophe)
* Please (capitalized)
* drive. (period)

File: extensions/sql-database-projects/src/common/utils.ts [hunk 2]
```diff
@@ -27,11 +28,22 @@ export function trimUri(innerUri: vscode.Uri, outerUri: vscode.Uri): string {
 	let innerParts = innerUri.path.split('/');
 	let outerParts = outerUri.path.split('/');
 
+	if (path.isAbsolute(outerUri.path)
+		&& innerParts.length > 0 && outerParts.length > 0
+		&& innerParts[0].toLowerCase() !== outerParts[0].toLowerCase()) {
+		throw new Error(constants.ousiderFolderPath);
```
Comment (Benjin) [inline extensions/sql-database-projects/src/common/utils.ts:34]: This error message explicitly mentions project file, but this util function gets used on more than just that, I think?  Maybe this specific error message be higher up (like in `readProjFile()`) after this line throws or this message should be more generic.

File: extensions/sql-database-projects/src/common/utils.ts [hunk 3]
```diff
@@ -27,11 +28,22 @@ export function trimUri(innerUri: vscode.Uri, outerUri: vscode.Uri): string {
 	let innerParts = innerUri.path.split('/');
 	let outerParts = outerUri.path.split('/');
 
+	if (path.isAbsolute(outerUri.path)
```
Comment (Benjin) [inline extensions/sql-database-projects/src/common/utils.ts:31]: update function's docstring with new unshift info/example

File: extensions/sql-database-projects/src/models/tree/fileFolderTreeItem.ts [hunk 4]
```diff
@@ -87,15 +88,18 @@ export function sortFileFolderNodes(a: (FolderNode | FileNode), b: (FolderNode |
...
 	if (fileSystemUri.fsPath.startsWith(projBaseDir)) {
 		localUri = fileSystemUri.fsPath.substring(projBaseDir.length);
 	}
-	else {
-		throw new Error(`Project (${projBaseDir}) pointing to file outside of directory (${fileSystemUri.fsPath})`);
+	else if (isFile) {
+		// if file is outside the folder add add at top level in tree
+		// this is not true for folders otherwise the outside files will not be directly inside the top level
+		let parts = utils.getPlatformSafeFileEntryPath(fileSystemUri.fsPath).split('/');
+		localUri = parts[parts.length - 1];
```
Comment (Benjin) [inline extensions/sql-database-projects/src/models/tree/fileFolderTreeItem.ts:102]: What's the intended behavior if I have a folder outside the project directory that contains files of its own?  Will the files be inside the folder that's located at the tree's top level, or will there be an empty folder alongside files all at the top level?

```
MyProj
    /OutsideFolder
        /OutsideFileInOutsideFolder.sql
```

or

```
MyProj
    /OutsideFolder
    /OutsideFileInOutsideFolder.sql
```

File: extensions/sql-database-projects/src/models/tree/fileFolderTreeItem.ts [hunk 5]
```diff
@@ -87,15 +88,18 @@ export function sortFileFolderNodes(a: (FolderNode | FileNode), b: (FolderNode |
...
 	const projBaseDir = projectNode.project.projectFolderPath;
 	let localUri = '';
 
 	if (fileSystemUri.fsPath.startsWith(projBaseDir)) {
 		localUri = fileSystemUri.fsPath.substring(projBaseDir.length);
 	}
-	else {
-		throw new Error(`Project (${projBaseDir}) pointing to file outside of directory (${fileSystemUri.fsPath})`);
+	else if (isFile) {
+		// if file is outside the folder add add at top level in tree
```
Comment (Benjin) [inline extensions/sql-database-projects/src/models/tree/fileFolderTreeItem.ts:99]: also, as I'm thinking more about this... should outside files/folders have some indication in the tree, or do we expect the user to not care as long as they're built and part of source control?

File: extensions/sql-database-projects/src/models/tree/projectTreeItem.ts [hunk 6]
```diff
@@ -51,6 +51,9 @@ export class ProjectRootTreeItem extends BaseProjectTreeItem {
 	 */
 	private construct() {
 		for (const entry of this.project.files) {
+			if (entry.type !== EntryType.File && entry.relativePath.startsWith(RelativeOuterPath)) {
+				continue;
+			}
```
Comment (Benjin) [inline extensions/sql-database-projects/src/models/tree/projectTreeItem.ts:56]: Nit: blank line after blocks

File: extensions/sql-database-projects/src/test/projectController.test.ts [hunk 7]
```diff
@@ -48,7 +48,13 @@ const mockConnectionProfile: azdata.IConnectionProfile = {
 
 
 
-describe ('ProjectsController', function(): void {
+describe('ProjectsController', function (): void {
+	// Moving to overall describe since it's used across multiple child functions
```
Comment (Benjin) [inline extensions/sql-database-projects/src/test/projectController.test.ts:52]: Code comment as opposed to PR comment?

File: extensions/sql-database-projects/src/test/projectTree.test.ts [hunk 8]
```diff
@@ -116,4 +116,22 @@ describe.skip('Project Tree tests', function (): void {
 			'MyNestedFolder2',
 			'MyFile2.sql']);
 	});
+
+	it('Should be able to parse and include relative paths outside project folder', function (): void {
+		const root = os.platform() === 'win32' ? 'Z:\\Level1\\Level2\\' : '/';
```
Comment (Benjin) [inline extensions/sql-database-projects/src/test/projectTree.test.ts:121]: There's no reason for any logic to check for this, but I find it funny that we're escaping beyond FS root.

File: extensions/sql-database-projects/src/test/utils.test.ts [hunk 9]
```diff
@@ -5,19 +5,39 @@
...
 		should(await exists(path.join(testFolderPath, 'folder2'))).equal(true);
 		should(await exists(path.join(testFolderPath, 'folder4'))).equal(false);
-		should(await exists(path.join(testFolderPath, 'folder2','file4.sql'))).equal(true);
-		should(await exists(path.join(testFolderPath, 'folder4','file2.sql'))).equal(false);
+		should(await exists(path.join(testFolderPath, 'folder2', 'file4.sql'))).equal(true);
+		should(await exists(path.join(testFolderPath, 'folder4', 'file2.sql'))).equal(false);
+	});
+
+	it('Should get correct relative paths of files/folders', async () => {
+		// test mac style paths
```
Comment (Benjin) [inline extensions/sql-database-projects/src/test/utils.test.ts:26]: comment no longer true

File: extensions/sql-database-projects/src/models/tree/fileFolderTreeItem.ts (hunk 4 above)
Comment (udeeshagautam) [reply extensions/sql-database-projects/src/models/tree/fileFolderTreeItem.ts:102]: MyProj
    /OutsideFileInOutsideFile.sql
    /OutsideFileInOutsideFile2.sql

(no outside folders in tree - as per SSDT)

File: extensions/sql-database-projects/src/models/tree/fileFolderTreeItem.ts (hunk 5 above)
Comment (udeeshagautam) [reply extensions/sql-database-projects/src/models/tree/fileFolderTreeItem.ts:99]: it good point - SSDT has different (link) icon

File: extensions/sql-database-projects/src/common/utils.ts (hunk 2 above)
Comment (udeeshagautam) [reply extensions/sql-database-projects/src/common/utils.ts:35]: I changed the msg a bit but also took a quick look at places where its used... most are with existing file operations so its okay to refer to project file as such.

**Pull request merged**